File: providers/apollo.py
"""Apollo provider for Mid-Market firmographic enrichment."""

import requests
from config import APOLLO_API_KEY

APOLLO_BASE_URL = "https://api.apollo.io/v1"


def enrich_company(domain: str = None, name: str = None) -> dict | None:
    """
    Enrich a company using Apollo's organization enrichment API.

    Args:
        domain: Company domain (e.g., "notion.com")
        name: Company name (fallback if domain not provided)

    Returns:
        Company data dict or None if not found
    """
    if not APOLLO_API_KEY:
        raise ValueError("APOLLO_API_KEY not configured")

    if not domain and not name:
        raise ValueError("Must provide either domain or name")

    payload = {"api_key": APOLLO_API_KEY}
    if domain:
        payload["domain"] = domain
    else:
        payload["organization_name"] = name

    response = requests.post(
        f"{APOLLO_BASE_URL}/organizations/enrich",
        json=payload,
        timeout=30,
    )
    response.raise_for_status()

    data = response.json()
    org = data.get("organization")

    if not org:
        return None

    return {
        "name": org.get("name", ""),
        "domain": org.get("primary_domain", ""),
        "industry": org.get("industry", ""),
        "employee_count": org.get("estimated_num_employees"),
        "revenue_range": _format_revenue(org.get("annual_revenue")),
        "founded_year": org.get("founded_year"),
        "linkedin_url": org.get("linkedin_url", ""),
        "description": org.get("short_description", ""),
        "city": org.get("city", ""),
        "state": org.get("state", ""),
        "country": org.get("country", ""),
    }


def search_contacts(
    domain: str,
    titles: list[str] = None,
    limit: int = 5,
) -> list[dict]:
    """
    Search for contacts at a company using Apollo's people search.

    Args:
        domain: Company domain
        titles: List of job titles to filter by (optional)
        limit: Max contacts to return

    Returns:
        List of contact dicts
    """
    if not APOLLO_API_KEY:
        raise ValueError("APOLLO_API_KEY not configured")

    payload = {
        "api_key": APOLLO_API_KEY,
        "q_organization_domains": domain,
        "per_page": limit,
    }

    if titles:
        payload["person_titles"] = titles

    response = requests.post(
        f"{APOLLO_BASE_URL}/mixed_people/search",
        json=payload,
        timeout=30,
    )
    response.raise_for_status()

    data = response.json()
    people = data.get("people", [])

    results = []
    for person in people:
        results.append({
            "name": person.get("name", ""),
            "title": person.get("title", ""),
            "email": person.get("email", ""),
            "linkedin_url": person.get("linkedin_url", ""),
            "city": person.get("city", ""),
            "state": person.get("state", ""),
        })

    return results


def _format_revenue(revenue: int | None) -> str:
    """Format revenue as readable string."""
    if not revenue:
        return "Unknown"
    if revenue >= 1_000_000_000:
        return f"${revenue / 1_000_000_000:.1f}B"
    if revenue >= 1_000_000:
        return f"${revenue / 1_000_000:.0f}M"
    if revenue >= 1_000:
        return f"${revenue / 1_000:.0f}K"
    return f"${revenue}"

File: frontend/lib/providers/apollo.ts
/**
 * Apollo Provider Adapter
 *
 * Mid-Market firmographic enrichment and B2B contact search.
 * Provides company data and verified contact information.
 *
 * API Docs: https://apolloio.github.io/apollo-api-docs/
 */

import {
  ProviderAdapter,
  ProviderResponse,
  ContactResponse,
  CompanyQuery,
  ContactQuery,
  ProviderError,
  createProviderResponse,
  createContactResponse,
  formatRevenue,
} from './types';

const APOLLO_BASE_URL = 'https://api.apollo.io/v1';

/**
 * Clean domain to format Apollo expects.
 * Removes protocol, www, paths, and whitespace.
 */
function cleanDomain(raw: string | null | undefined): string | null {
  if (!raw) return null;
  return raw
    .replace(/^https?:\/\//, '')  // remove protocol
    .replace(/\/.*$/, '')          // remove path
    .replace(/^www\./, '')         // remove www
    .toLowerCase()
    .trim();
}

/**
 * Apollo Provider Adapter
 *
 * Provides company enrichment and contact search capabilities.
 */
export class ApolloAdapter implements ProviderAdapter {
  id = 'apollo';
  name = 'Apollo';
  private apiKey?: string;

  constructor(apiKey?: string) {
    this.apiKey = apiKey;
  }

  /**
   * Get Apollo API key from constructor (must be passed from provider_connections).
   */
  private getApiKey(): string {
    if (!this.apiKey) {
      throw new ProviderError('apollo', 'config_error', 'Apollo API key not provided. Must be retrieved from provider_connections table.');
    }
    return this.apiKey;
  }

  /**
   * Enrich a company using Apollo's organization enrichment API.
   */
  async enrichCompany(query: CompanyQuery): Promise<ProviderResponse | null> {
    const apiKey = this.getApiKey();

    // Clean domain before sending to Apollo
    const cleanedDomain = cleanDomain(query.domain);

    if (!cleanedDomain && !query.name) {
      console.log(`[Apollo] Skipping record: no valid domain after cleaning`);
      return null;
    }

    const payload: Record<string, string> = {
      api_key: apiKey,
    };

    if (cleanedDomain) {
      payload.domain = cleanedDomain;
    } else if (query.name) {
      payload.organization_name = query.name;
    }

    const response = await fetch(`${APOLLO_BASE_URL}/organizations/enrich`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      throw new ProviderError(
        'apollo',
        'api_error',
        `API request failed: ${response.statusText}`,
        response.status
      );
    }

    const data = await response.json();
    const org = data.organization;

    if (!org) {
      return null;
    }

    return createProviderResponse(
      this.id,
      org,
      {
        name: org.name || '',
        domain: org.primary_domain || '',
        industry: org.industry || '',
        employee_count: org.estimated_num_employees,
        revenue_range: formatRevenue(org.annual_revenue),
        founded_year: org.founded_year,
        linkedin_url: org.linkedin_url || '',
        description: org.short_description || '',
        city: org.city || '',
        state: org.state || '',
        country: org.country || '',
      }
    );
  }

  /**
   * Search for contacts at a company using Apollo's people search.
   */
  async searchContacts(query: ContactQuery): Promise<ContactResponse[]> {
    const apiKey = this.getApiKey();

    const payload: Record<string, unknown> = {
      api_key: apiKey,
      q_organization_domains: query.domain,
      per_page: query.limit || 5,
    };

    if (query.titles && query.titles.length > 0) {
      payload.person_titles = query.titles;
    }

    const response = await fetch(`${APOLLO_BASE_URL}/mixed_people/search`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      throw new ProviderError(
        'apollo',
        'api_error',
        `API request failed: ${response.statusText}`,
        response.status
      );
    }

    const data = await response.json();
    const people = data.people || [];

    return people.map((person: Record<string, unknown>) =>
      createContactResponse(this.id, {
        name: (person.name as string) || '',
        title: (person.title as string) || '',
        email: (person.email as string) || '',
        linkedin_url: (person.linkedin_url as string) || '',
        city: (person.city as string) || '',
        state: (person.state as string) || '',
        raw: person,
      })
    );
  }
}

export default ApolloAdapter;
